src/gc.c

```diff
@@ -74,31 +73,26 @@ static void threadCallback(void* data) {
       RedisModule_UnblockClient(bc, NULL);
     }
     rm_free(task);
-    goto end;
-  }
-
-  if (!ret) {
+  } else if (!ret) {
     // The index was freed. There is no need to reschedule the task.
     // We need to free the task and the GC.
     RedisModule_Log(RSDummyContext, REDISMODULE_LOGLEVEL_DEBUG, "GC %p: Self-Terminating. Index was freed.", gc);
     gc->callbacks.onTerm(gc->gcCtx);
     if (bc) RedisModule_UnblockClient(bc, NULL);
```

Comment:
I think `if (bc) RedisModule_UnblockClient(bc, NULL);` should be outside any if.
also, we can have a local flag `free_task` which defaults to 1 and is turned off only `if (gc->timerID)`
Then we don't need to check if it's a debug task (unless there is some risk in locking in debug ?), or at least use it only to skip to end and make this if block less complicated.

Reply:
Addressing the main comment:

1. Not that I'm aware of. Not sure what can leak there
2. Almost all. The inner case of the `else` section is timing-related, so there is no way of testing it consistently. This is also a debug-command-related only, so no real-life scenario of reaching this case

Addressing comment:
1. I agree that currently it can be moved outside, but usually, there shouldn't be a blocked client so I prefer to have this condition only when we know we might have a blocked client. This way we also have a better track of file history (fewer lines modified
2. Here again, in the common case we don't free the task. only on force-invocation from a debug command (the task is allocated for these commands, separately from the periodic task), or if the GC was requested to stop scheduling tasks (also from a debug command). In my opinion, we should have as few as possible checks in the common flow, so it is better to free the task only if we need to, even if it means calling `rm_free` from a few different places.

In the common case:
- `ret == 1` (unless the index was dropped - happens once per index in production)
- `task->debug == false` (only true in force invocation - not in production)
- `gc->timerID != 0` (only 0 if it was stopped by a debug command - no in production)
- `bc == NULL` only non-NULL on the blocking force invocation

Reply:
Isn't it a leak to not call gc->callbacks.onTerm(gc->gcCtx); and rm_free(gc); (on debug) ?

regarding the comment: So to emphasize what is the most common scenario i would handle `ret == true` first (also less ifs...)

something like
if (ret) { //scedulenext if needed //return}

//free stuff

what do you think?

Reply:
I’ll rearrange the cases so the first one is the most common one.
It is not a leak because we want to call `onTerm` and free the GC only if `ret == 0`